File: practices/interview-bit/stacks-and-queues/nearest-smaller-element.cpp
// O(N) in space
// O(2N) = O(N) in time
vector<int> Solution::prevSmaller(vector<int> &A) {
    stack<int> s;
    vector<int> ans;
    for(int i = 0; i < (int)A.size(); i++) {
        while(!s.empty() && s.top() >= A[i]) {
            s.pop();
        }

        ans.push_back(s.empty() ? -1 : s.top());
        s.push(A[i]);
    }

    return ans;
}

File: practices/interview-bit/stacks-and-queues/nearest-smaller-element.py
class Solution:
    # @param A: list of integers
    # @return a list of integers
    def prevSmaller(self, A):
        s = []
        ans = []
        for x in A:
            while s and s[-1] >= x:
                s.pop()

            ans.append(-1 if not s else s[-1])
            s.append(x)

        return ans
